### ISSUE TITLE
Unable to add or update user

### ISSUE THREAD

**Issue by kichetof** [Adldap2/Adldap2#90]
Hi Steve,

I encountered a bug when I want to add or update a user, this is a part of my code:

```php
//[...] connection, etc
if($_edit){ // Edition mode, find DN from AD
    $user = $adldap->search()->findByDn($attributes['dinstinguishedName']);
    $log->addInfo("DN :".$user->getDn());
}

if($_add){ // Adding mode, create new instance
    $user = $adldap->users()->newInstance();
}
//[...]
if($_add){
    $user->setDn($attributes['dinstinguishedName']);
}
if($_edit && $user->getDn() != $attributes['dinstinguishedName']){
    $user->move("CN={$attributes["cn"]}", "OU=Users,OU=$userOU,".BASE_DN);
}
$user->setFirstName($attributes['givenname']);
$user->setLastName($attributes['sn']);

$user->setName($attributes['sn']);
$user->setCommonName($attributes["displayname"]);
$user->setDisplayName($attributes['displayname']);

if($user->save()){ // BUG
    //[...]
}
```
__Add error:__ ` PHP Warning:  ldap_add(): Add: Invalid syntax in vendor/adldap2/adldap2/src/Connections/Ldap.php on line 520`

__Edit error:__ `PHP Warning:  ldap_modify_batch(): Batch Modify: Operation not allowed on RDN in vendor/adldap2/adldap2/src/Connections/Ldap.php on line 589`

Please find the users object before `$user->save()`
http://pastebin.com/zbMsEkhQ

I don't know what changes affect my code ?
Thanks for your help!
Tof

**Comment by stevebauman** [comment]
Hi @kichetof, the DN appears to be wrong on line 32 of your pastebin. You're setting the users DN as the base DN of your LDAP server. It should appear:

```php
'dn' => 'cn=John Doe,dc=NAME,dc=DOMAIN,dc=local',
```

*Not*

```php
'dn' => 'dc=NAME,dc=DOMAIN,dc=local',
```

Your DN is not built /modified automatically when creating or modifying a user. A lot of assumption would have to take place for this to happen.

**Comment by kichetof** [comment]
Hi Steve,
Thanks for your reply. The DN on line 32 correspond of the base DN setted on my AD connection

```php
$config = new \Adldap\Connections\Configuration();

$config->setAccountSuffix(ACCOUNT_SUFFIX);
$config->setDomainControllers(unserialize(DOMAIN_CONTROLLERS));
$config->setBaseDn(BASE_DN);
$config->setAdminUsername(ADMIN_USERNAME);
$config->setAdminPassword(ADMIN_PASSWORD);
$config->setFollowReferrals(false);
$config->setUseSSL(true);

$adldap = new adLDAP($config);
````

On line 115, the DN is `'dn' => 'CN=Testcf300901 TESTCF300901,OU=Users,DC=NAME,DC=DOMAIN,DC=local',`

It seems that the DN values isn't an array

```php
'dn' => 'CN=Testcf300901 TESTCF300901,OU=Users,DC=NAME,DC=DOMAIN,DC=local',
```

And maybe need to be

```php
'dn' =>
    array (
        0 => 'CN=Testcf300901 TESTCF300901,OU=Users,DC=NAME,DC=DOMAIN,DC=local',
    ),
```

What do you think ?

I always use this code to add user _(I only added $_edit mode today)_

**Comment by stevebauman** [comment]
Whoops you're right, didn't look further into the paste bin, apologies!

The DN attribute doesn't need to be an array, as it's not actually passed as an attribute to LDAP. It's passed separately into the `ldap_add()` method: https://github.com/Adldap2/Adldap2/blob/master/src/Models/AbstractModel.php#L819

The invalid syntax error would indicate something is wrong with your DN, or you may be setting a user attribute that must be either a string / int. I'll keep looking, but try adding a plain user first and see if you're able to do that.

**Comment by kichetof** [comment]
Okay I found the error, at least I know where it comes from...

`$user->setLocale($attributes['l']);`

but... that works on my last code and not now, I'll investigate tomorrow!
Thanks for your help!
I'll keep this issue open until I found and share the solution :)

**Comment by kichetof** [comment]
__Solution found !!__

I replaced all my var that contain nothing with "", but AD doesn't like that!

you need to set an empty value to " " // <- note the space between quotes!

`$var = isset($value) ? $value : " ";`

**Comment by kichetof** [comment]
Hi @stevebauman,

New error with update :(

`PHP Warning:  ldap_modify_batch(): Each element of a 'values' array must be a string in /var/www/html/tools/AD/vendor/adldap2/adldap2/src/Connections/Ldap.php on line 589`

It seems that the problem come from 2 attributes: __objectsid__ & __objectguid__

```php
["objectguid"]=>
    array(1) {
      [0]=>
      string(16) "�w��
M�ٹ�Y��"
    }
    [24]=>
    string(10) "objectguid"
["objectsid"]=>
    array(1) {
      [0]=>
      string(28) ".d
9.���
"
```

As you can see, there are a lot of ? on the attributes.
I have a same issue with Monolog when I try to export var
```php
PHP Fatal error:  Uncaught exception 'RuntimeException' with message 'JSON encoding failed: Malformed UTF-8 characters, possibly incorrectly encoded. Encoding: array (\n  0 => \n  array (\n    'Type' => 'INFO',\n    'File' => '',\n    'Line' => '',\n    'Label' => 'Adduser',\n  ),\n  1 => 'L\\'object user:Adldap\\\\Models\\\\User::__set_state(array(\n   \\'exists\\' => true,\n   \\'dateFormat\\' => \\'Y-m-d H:i:s\\',\n   \\'query\\' => \n  Adldap\\\\Query\\\\Builder::__set_state(array(\n     \\'bindings\\' => \n    array (\n      \\'where\\' => \\'wheres\\',\n      \\'orWhere\\' => \\'orWheres\\',\n    ),\n     \\'selects\\' => \n    array (\n    ),\n     \\'wheres\\' => \n    array (\n      0 => \n      array (\n        \\'field\\' => \\'objectclass\\',\n        \\'operator\\' => \\'*\\',\n        \\'value\\' => \\'\\',\n      ),\n    ),\n     \\'orWheres\\' => \n    array (\n    ),\n     \\'filters\\' => \n    array (\n    ),\n     \\'paginated\\' => false,\n     \\'sortByField\\' => \\'\\',\n     \\'sortByDirection\\' => \\'\\',\n     \\'dn\\' => \\'CN=Testcf021002 TESTCF021002,OU=Users,DC=NAME,DC=UN in /var/www/html/tools/AD/vendor/monolog/monolog/src/Monolog/Formatter/NormalizerFormatter.php on line 191,
```

This problem appear only when I try to update an existing user.

I hope you'll have a solution :)
Big thanks!
Tof

**Comment by stevebauman** [comment]
That's strange, those attributes shouldn't be an issue unless you're modifying them. Only attributes that are modified are sent to active directory. Are you explicitly setting anything to an integer or modifying the objectSID or the objectGUID?

**Comment by kichetof** [comment]
No setting for ObjectSID/GUID
The only integer setter come from Utilities::ConvertUnixTimeToWindowsTime all my setter return a string.
The strange thing is that works when I insert a new user and crash when I try to edit one...

**Comment by stevebauman** [comment]
Which attribute are you setting that involves `Utilities::ConvertUnixTimeToWindowsTime()`?

**Comment by kichetof** [comment]
If I remember well is useraccountexpires or something like that :)

**Comment by stevebauman** [comment]
If you don't modify the account expires attribute, are you able to save successfully?

**Comment by kichetof** [comment]
I'll try it monday :)

**Comment by stevebauman** [comment]
Ok, I'm assuming this is the issue. I think you'll have to cast the integer to a string when you set it using

```php
// Note the (string) before the utilities call
$accountExpires = (string) Utilities::ConvertUnixTimeToWindowsTime(1443803939);
```

**Comment by kichetof** [comment]
Thanks for the tips! That works fine!

Now I'm trying to solve a new bug

`ldap_modify_batch(): Batch Modify: Type or value exists in vendor/adldap2/adldap2/src/Connections/Ldap.php on line 589`

That the explication of the error: 
> «Type or value exists» can be either you are trying to insert twice the same values in a multi valued field or you are trying to insert several values in a single valued field
http://stackoverflow.com/a/16915425

Here are the attributes

```php
array(45) {
    ["objectclass"]=> array(4) {
        [0]=> string(3) "top"
        [1]=> string(6) "person"
        [2]=> string(20) "organizationalPerson"
        [3]=> string(4) "user"
    }
    [0]=> string(11) "objectclass"
    ["sn"]=> array(1) {
        [0]=> string(8) "DOE"
    }
    [1]=> string(2) "sn"
    ["c"]=> array(1) {
        [0]=> string(2) "CH"
    }
    [2]=> string(1) "c"
    ["l"]=> array(1) {
        [0]=> string(7) "Genève"
    }
    [3]=> string(1) "l"
    ["title"]=> array(1) {
        [0]=> string(1) " "
    }
    [4]=> string(5) "title"
    ["description"]=> array(1) {
        [0]=> string(1) " "
    }
    [5]=> string(11) "description"
    ["postalcode"]=> array(1) {
        [0]=> string(4) "1200"
    }
    [6]=> string(10) "postalcode"
    ["physicaldeliveryofficename"]=> array(1) {
        [0]=> string(17) "Bat I;0001"
    }
    [7]=> string(26) "physicaldeliveryofficename"
    ["telephonenumber"]=> array(1) {
        [0]=> string(15) "123456789"
    }
    [8]=> string(15) "telephonenumber"
    ["givenname"]=> array(1) {
        [0]=> string(5) "John"
    }
    [9]=> string(9) "givenname"
    ["initials"]=> array(1) {
        [0]=> string(2) "JD"
    }
    [10]=> string(8) "initials"
    ["displayname"]=> array(1) {
        [0]=> string(14) "John DOE"
    }
    [11]=> string(11) "displayname"
    ["department"]=> array(1) {
        [0]=> string(9) "TEST"
    }
    [12]=> string(10) "department"
    ["company"]=> array(1) {
        [0]=> string(22) "My company"
    }
    [13]=> string(7) "company"
    ["streetaddress"]=> array(1) {
        [0]=> string(47) "Bat 1
streetaddress"
    }
    [14]=> string(13) "streetaddress"
    ["samaccountname"]=> array(1) {
        [0]=> string(8) "doej"
    }
    [15]=> string(14) "samaccountname"
    ["userprincipalname"]=> array(1) {
        [0]=> string(25) "doej@NAME.DOMAIN.local"
    }
    [16]=> string(17) "userprincipalname"
    ["objectcategory"]=> array(1) {
        [0]=> string(62) "CN=Person,CN=Schema,CN=Configuration,DC=NAME,DC=DOMAIN,DC=local"
    }
    [17]=> string(14) "objectcategory"
    ["mail"]=> array(1) {
        [0]=> string(23) "John.DOE@domain.local"
    }
    [18]=> string(4) "mail"
    ["manager"]=> array(1) {
        [0]=> string(76) "CN=Administrator,OU=Users,DC=NAME,DC=DOMAIN,DC=local"
    }
    [19]=> string(7) "manager"
    ["dn"]=> string(73) "CN=John DOE,OU=Users,DC=NAME,DC=DOMAIN,DC=local"
    ["name"]=> array(1) {
        [0]=> string(8) "DOE"
    }
    ["cn"]=> array(1) {
        [0]=> string(14) "John DOE"
    }
    ["info"]=> array(1) {
        [0]=> string(1) " "
    }
    ["accountexpires"]=> array(1) {
        [0]=> string(18) "131038488000000000"
    }
}
```

It seems that is missing something like that :

```php
    [20]=> string(4) "name"

    [21]=> string(2) "cn"

    [22]=> string(4) "info"

    [23]=> string(14) "accountexpires"

**Comment by stevebauman** [comment]
Apologies for long response. Unfortunately I need to see the modified attributes array. Can you call the method `$user->getModifications()` (after you've modified the object of course) and paste the array? This would let me know what's being pushed to AD so I can help you troubleshoot further.

**Comment by kichetof** [comment]
Hi @stevebauman, don't worry we're all busy :)
Please find `$user->getModifications()`

```php
// MODIFICATIONS
array(6) {
    [0]=> array(3) {
    ["attrib"]=> string(5) "title"
        ["modtype"]=> int(3)
        ["values"]=> array(1) {
            [0]=> string(1) " "
        }
    }
    [1]=> array(3) {
    ["attrib"]=> string(11) "description"
        ["modtype"]=> int(3)
        ["values"]=> array(1) {
            [0]=> string(1) " "
        }
    }
    [4]=> array(3) {
    ["attrib"]=> string(4) "info"
        ["modtype"]=> int(1)
        ["values"]=> array(1) {
            [0]=> string(1) " "
        }
    }
    [5]=> array(3) {
    ["attrib"]=> string(14) "accountexpires"
        ["modtype"]=> int(1)
        ["values"]=> array(1) {
            [0]=>     string(18) "131038488000000000"
        }
    }
}
```

Thanks for your help!

**Comment by stevebauman** [comment]
Strange... have you tried modifying and saving the user without setting the account expires?

**Comment by kichetof** [comment]
Yes... And try with null or " " --> error

**Comment by stevebauman** [comment]
I'm really not sure here, everything seems fine and I've tried modifying the same fields you have listed here and I'm able to save without issue. I'm going to need some time to look into it.

**Comment by kichetof** [comment]
I've found some error on my AD:

#### Setter AccountExpires
```
Internal event: The LDAP server returned an error. 
 
Additional Data 
Error value:
0000207E: AtrErr: DSID-03190647, #1:
	0: 0000207E: DSID-03190647, problem 1006 (ATT_OR_VALUE_EXISTS), data 0, Att 9009f (accountExpires)
```
_(setters AccountExpires was disable to obtain this error)_
#### Setters Info
```
Internal event: The LDAP server returned an error. 
 
Additional Data 
Error value:
00002083: AtrErr: DSID-031513D7, #1:
	0: 00002083: DSID-031513D7, problem 1006 (ATT_OR_VALUE_EXISTS), data 0, Att 20051 (info):len 2
```

I hope you could find a way to resolve this error :)

**Comment by stevebauman** [comment]
That's strange, if you're inserting a string into the `setInfo()` method then a string will be pushed to AD. No conversions should be done before hand. I'm able to update a users info without issue.

Also looking into the account expires integer.

**Comment by kichetof** [comment]
Thanks for your commits! :)

I encounter the same error with EmployeeID when I want to edit them!
If I empty the field from AD, I'm able to edit it with my code.
If the field is filled, error :(


```php
var_export($user->getDirty(), true);

array (
    'title' => array (
        0 => NULL,
    ),
    'info' => array (
        0 => NULL,
    ),
    'description' => array (
        0 => NULL,
    ),
    'telephonenumber' => array (
        0 => NULL,
    ),
    'employeeid' => array (
        0 => '0124123',
    ),
)
```
``` 
Internal event: The LDAP server returned an error. 
 
Additional Data 
Error value:
00002081: AtrErr: DSID-030F154F, #1:
	0: 00002081: DSID-030F154F, problem 1006 (ATT_OR_VALUE_EXISTS), data 0, Att 90023 (employeeID)
```

**Comment by stevebauman** [comment]
I'm really not sure. I'm able to edit and save these attributes fine through my AD server. Researching your AD error seems like it may be a synchronization issue if you have any other programs hooked into your AD server.

**Comment by kichetof** [comment]
# FIXED!!!

```php
define("FIELDS_INFO_USER",
   serialize(
       [
           'samaccountname',
           'userprincipalname',
           'dinstinguishedName',
           'sn',
           'givenName',
           'mail',
           'description',
           'physicaldeliveryofficename',
           'l',
           'postalCode',
           'c',
           'streetAddress',
           'company',
           'department',
           'displayName',
           'initials',
           'title',
           'ipphone',
           'telephoneNumber',
           'info',
           'description',
           'userDateExpiry',
           'userDateEntry',
           'manager',
           'cn',
           'name',
           'accountexpires',
           'employeeid',
       ]
   )
);

$user = $adldap->search()->findByDn($attributes['dinstinguishedName'], unserialize(FIELDS_INFO_USER));

$attributes["xxx"] = isset($_POST['xxx']) ? $_POST['xxx'] : NULL;
$user->setXxx($attributes["xxx"]);
$user->save();
```

If an attributes isn't selected when we `findByDn` we get this error!
Plus my fix to reset the AccountExpires to _9223372036854775807_

Many thanks for your great support!

**Comment by stevebauman** [comment]
Glad to hear you've resolved it, thanks for the pull-request!
